### ISSUE TITLE
Android Rpush Notification doesnt get called when app is in background after set content_available

### ISSUE THREAD

**Issue by boyfunky** [rpush/rpush#260]
So this is my code to get my notification event to be fired when the app is in background on the client side but it doesnt work. 

```
n = Rpush::Gcm::Notification.new
n.app = Rpush::Gcm::App.find_by_name("partners")
n.registration_ids = [@user_auth_token.device_token]
n.data = { state: "case_status_single", count: badge_unread.length + 1, notification_id: n.id, content_available: '1'}
n.content_available = true
n.notification = { 
    body: @loan_application.name + " Loan Status Changed to: " + AgentAssignmentStatus.find(p.agent_assignment_status_id).assignment_status,
    title: Partners',
    tag: true
 }
 n.save!
 Rpush.push
```

I thought with the content_available parameter allows the onNotification event from the mobile app side to be triggered even if the app is in background. 

Any thoughts on this?

Thanks

**Comment by aried3r** [comment]
`content_available` is only for iOS currently. On Android, the data message wakes up the device anyway. There is however the `priority` which can be either `normal` or `high`. `high` should wake up even devices with Doze enabled. Note: You usually do not want to send all messages as high as it will always wake up the device and would abuse the service.

For more about the priority, read more here:
https://developers.google.com/cloud-messaging/concept-options#setting-the-priority-of-a-message

Also the quote of Google Engineer, Dianne Hackborn:
> Abuse of high priority messages have a special difference from other things like notifications: they must go through Google servers, so Google can monitor and modify what is being sent to devices.  If apps abuse these for other things besides their intended use, we will be able to stop that abuse without touching any software on the device.  (Also "abuse" here is much less subjective than for notifications, where there is a large gray area of things some users care about and some don't.  For high priority messages, if it isn't something that is time critical to go to the user immediately, it is not appropriate.)

https://plus.google.com/105051985738280261832/posts/4x9tJoFWNof

**Comment by boyfunky** [comment]
@aried3r yes i understand that android should wake up the device but it doesnt. It only catches the onNotification event when the app is in the foreground and doesnt wake it when it is in the background. So how can i handle this?

**Comment by aried3r** [comment]
I don't think this is a problem with rpush but your Android app. You could try to verify by manually sending the same payload via Postman.
We use the same kind of payload for our app at work, so I don't think it's that.

**Comment by fertobar** [comment]
hello, I notice I set  priority **high** for GCM, but on Payload I see **priority=>10** , that is ok for GCM?

according to doc, seems priority 10 is for APN and high/normal for GCM.
or I'm missing something?
thanks in advance

**Comment by fertobar** [comment]
@aried3r , I think the issue could be here:
https://github.com/rpush/rpush/blob/644ba8fd5372aa974eb42c1db42a1ebf8ed35604/lib/rpush/client/active_model/gcm/notification.rb#L6

GCM priority is defined in base to the APN values.
For APN is 10 or 5, for GCM is not a number, I will make some tries and propose a fix

**Comment by aried3r** [comment]
> or I'm missing something?

Admittedly the code might seem confusing. `priority` was first defined for APNS as an integer. Thus it could not be directly used for GCM/FCM. But as you can see [in the code (comment)](https://github.com/rpush/rpush/blob/e398be38ba7ae1f38000be5908e642ac3ac69b48/lib/rpush/client/active_model/gcm/notification.rb#L22-L35) there is a workaround (or "hack").

There's also [specs](https://github.com/rpush/rpush/blob/e398be38ba7ae1f38000be5908e642ac3ac69b48/spec/unit/client/active_record/gcm/notification_spec.rb#L40-L57) you can read to see how it behaves.

tl;dr: It works as you would expect when using `high` and `normal` for FCM.

**Comment by fertobar** [comment]
I notice an issue on GCM with priority
It's ok, I create with 'high'  and I see internally is converted to 10.

but then I see there is case statement:
https://github.com/rpush/rpush/blob/644ba8fd5372aa974eb42c1db42a1ebf8ed35604/lib/rpush/client/active_model/gcm/notification.rb#L27

if I log the value before the case , I see as 10, so there is no match and I see the condition:

'must be one of either "normal" or "high"'

after that set, priority is missing and never sent to FCM.
I added more log on the FCM post and I could confirm is not there in the payload

**Comment by aried3r** [comment]
Can you write a test case that fails in the way you describe? You can find a link to the existing test cases in my previous post.

**Comment by fertobar** [comment]
sure, I can also provide a potential fix for this issue

**Comment by aried3r** [comment]
That would be great!

**Comment by fertobar** [comment]
@aried3r seems hard to reproduce with specs, but something like this:

```
 def priority=(priority)
            case priority
              when 'high',10
                super(GCM_PRIORITY_HIGH)
              when 'normal',5
                super(GCM_PRIORITY_NORMAL)
              else
                errors.add(:priority, 'must be one of either "normal" or "high"')
            end
          end
on 
```https://github.com/rpush/rpush/blob/644ba8fd5372aa974eb42c1db42a1ebf8ed35604/lib/rpush/client/active_model/gcm/notification.rb#L28

worked for me.
May be special flow or condition that is generating the condition that I see the logs

**Comment by aried3r** [comment]
Yes, it's probably the way you log things. Do you have this problem in production? Because we use this feature successfully in our code.

**Comment by fertobar** [comment]
yes, actually the issue was reported by Android team to me and I could confirm after add more log that they are right. I could detect this adding log in the last post to FCM:
https://github.com/rpush/rpush/blob/ee32bdf3456f00d13d7c8ade28a6f6bf24257cb0/lib/rpush/daemon/gcm/delivery.rb#L147 and the priority field was missing there.

Also checked with extra log that  value was 10 so in the flow the else condition was selected there.

I was using version 3.0.0 and message created with **create** method, not new & save

**Comment by aried3r** [comment]
Could you post your code you use to build the notification?

**Comment by fertobar** [comment]
is like this:
```
    Rpush::Gcm::Notification.create({
      app:  app,
      ....
      data: data,
      priority: 'high',
      ....
    })
```

**Comment by fertobar** [comment]
this fixed the issue in my case: https://github.com/rpush/rpush/pull/353
feel free to customize or suggest other alternative

**Comment by aried3r** [comment]
Using your code I had no problem:
```Ruby
notification = Rpush::Gcm::Notification.create(
    app: app,
    priority: 'high',
    registration_ids: ['invalid'])

notification.as_json
=> {
  "registration_ids" => [
    [0] "invalid"
  ],
  "delay_while_idle" => false,
              "data" => nil,
          "priority" => "high",
      "time_to_live" => 86400
}

notification.as_json.to_json
=> "{\"registration_ids\":[\"invalid\"],\"delay_while_idle\":false,\"data\":null,\"priority\":\"high\",\"time_to_live\":86400}"
```

Before we merge #353 I'd love to see a failing test case that is fixed by your PR.

**Comment by fertobar** [comment]
great! other detail is that I'm on Redis storage.
I will try to reproduce with original code

**Comment by aried3r** [comment]
Closing, I could not reproduce this issue with PostgreSQL for persistence, maybe it's only for redis, but unless we get a reproducible error case for this, I'm in favor for leaving the code as it is.

I can vouch that it works as intended for FCM for both Android and iOS.

Should you think this is still an issue, please leave a short reply. :)

**Comment by aross** [comment]
This is definitely still a problem for FCM and redis store.  I haven't really dug down into what's going on, but it looks like a validation or something in Modis is calling with the priority setter again and since the value is 10 instead of 'high' it's triggering the error and setting to nil.  @fertobar 's solution works because it also allows a value of 10 or 5 in the setter.

**Comment by daniel-nelson** [comment]
I've tracked down the issue when using Redis:

1. Priority is set during notification creation: `n.priority = 'normal'`
2. The [priority setter](https://github.com/rpush/rpush/blob/master/lib/rpush/client/active_model/gcm/notification.rb#L23-L36) converts this to 5
3. The 5 is stored in Redis
4. When the notification is hydrated during `Rpush.push`, Modis [assign_attributes](https://github.com/rpush/modis/blob/master/lib/modis/attribute.rb#L78-L83) calls the `priority=` setter, passing it the 5 from Redis. Since 5 is neither 'normal' nor 'high' in the priority setter it reaches the (else case)[https://github.com/rpush/rpush/blob/master/lib/rpush/client/active_model/gcm/notification.rb#L34], setting an error on the hydrated instance

The [PR](https://github.com/rpush/rpush/pull/353) provided by @fertobar seems like the simplest fix for this.

**Comment by aried3r** [comment]
Thank you for looking into this, @daniel-nelson. Do you think you can write a failing test case that shows the faulty behavior and is fixed by the proposed code in the PR?

Reopening for now.

**Comment by aried3r** [comment]
Thanks for everyone's work and sorry for addressing this so late!

If you think the documentation is now lacking information, PRs would be more than welcome.
